surfactant/cmd/cli.py

```diff
@@ -52,16 +55,135 @@ def find(sbom, output_format, input_format, **kwargs):
 @click.command("edit")
-def edit(sbom):
+def edit(sbom, output_format, input_format, **kwargs):
     "CLI command to edit specific entry(s) in a supplied SBOM"
 
 
-@click.argument("sbom", type=click.File("r"), required=True)
-@click.command("add")
-def add(sbom):
-    "CLI command to add specific entry(s) to a supplied SBOM"
+class cli_add:
+    """
+    A class that implements the surfactant cli find functionality
+
+    Attributes:
+    match_functions     A dictionary of functions that provide matching functionality for given SBOM fields (i.e. uuid, sha256, installpath, etc)
+    sbom                An internal record of sbom entries the class adds to as it finds more matches.
+    """
+
+    camel_case_conversions: dict
+    match_functions: dict
+    sbom: SBOM
+
+    def __init__(self):
+        """Initializes the cli_add class"""
+        self.match_functions = {
+            "relationship": self.add_relationship,
+            "file": self.add_file,
+            "installPath": self.add_installpath,
+            "containerPath": self.add_containerpath,
+            "entry": self.add_entry,
+        }
+        self.camel_case_conversions = {
+            "uuid": "UUID",
+            "filename": "fileName",
+            "containerpath": "containerPath",
+            "installpath": "installPath",
+            "capturetime": "captureTime",
+            "relationshipassertion": "relationshipAssertion",
+        }
+
+    def handle_kwargs(self, kwargs: dict) -> dict:
+        converted_kwargs = {}
+        for k, v in kwargs.items():  # Convert key values to camelcase where appropriate
+            key = self.camel_case_conversions[k] if k in self.camel_case_conversions else k
+            converted_kwargs[key] = v
+        return converted_kwargs
+
+    def execute(self, input_sbom: SBOM, **kwargs):
+        """Executes the main functionality of the cli_find class
+        param: input_sbom   The sbom to find matches within
+        param: kwargs:      Dictionary of key/value pairs indicating what features to match on
+        """
+        converted_kwargs = self.handle_kwargs(kwargs)
+        self.sbom = input_sbom
+
+        for key, value in converted_kwargs.items():
+            if key in self.match_functions:
+                self.match_functions[key](value)
+            else:
+                logger.warning(f"Paramter {key} is not supported")
+        return self.sbom
+
+    def add_relationship(self, value: dict) -> bool:
+        self.sbom.add_relationship(Relationship(**value))
+
+    def add_file(self, path):
+        self.sbom.software.append(Software.create_software_from_file(path))
+
+    def add_entry(self, entry):
+        self.sbom.software.append(Software.from_dict(entry))
+
+    def add_installpath(self, path):
+        for sw in self.sbom.software:
+            sw.installPath.append(path)
+
+    def add_containerpath(self, path):
+        for sw in self.sbom.software:
+            sw.containerPath.append(path)
```

Comment:
I don't think these will work as expecte  since the CyTRICS SBOM schema uses paths that includes the file name at the end (e.g. `/usr/local/bin/mybinary`, which helps distinguish the install/container paths for multiple files that have different names installed to different locations (or symlinks).

My naive take would be that the provided install/container path should be a "prefix" that gets prepended to every file name associated with each software entry, but that would result in paths that just look like every single file is in the same directory. The problem I see is that the SBOM schema(s) don't preserve enough information on what subdirectory each file should be in relative to an install/container path prefix (e.g. the middle part of `<install/container_prefix><subdirectories><filename>` ).

Maybe this functionality should require supplying a "original prefix", then any time a container/install path has that prefix, add a new entry that has it swapped out for a new prefix?

Reply:
Ah I see that is a good point. I think your suggestion is a good option for a first pass, but I do think it is easier to just find and replace within a code editor than to do it via command line. Would there be a benefit to having this functionality as a cli option? I guess if someone wanted to script edits using surfactant, I can see them using this instead of VSCode or another IDE

Reply:
Args to cli command: abc0/ -> /usr/local
containerpath (abc0/bin/helics_binary) -> installpath (/usr/local/bin/helics_binary)